pkg/lang/ir/v1/system.go

```diff
@@ -46,6 +47,95 @@ func (g generalGraph) compileUbuntuAPT(root llb.State) llb.State {
+
+	aptSign := root.
+		File(llb.Mkdir(filePath, 0755, llb.WithParents(true)),
+			llb.WithCustomName("[internal] setting target apt-source signature folder")).
+		File(llb.Copy(builder, fileName, filePath+fileName),
+			llb.WithCustomName("[internal] copy signature from builder"))
+
+	return aptSign, filePath + fileName
+}
+
+func (g generalGraph) configRSrc(root llb.State, aptConfig ir.APTConfig, sign string) (llb.State, string) {
+
+	// Created different fields in each line to form up the .sources file
+	// The fields follow the DEB822 format spec
+
+	// "Enabled" field
+	var enabled = fmt.Sprintf("Enabled: %s\n", aptConfig.Enabled)
+	// "Types" field
+	var types = fmt.Sprintf("Types: %s\n", aptConfig.Types)
+	// "URIs" field
+	var uris = fmt.Sprintf("URIs: %s\n", aptConfig.URIs)
+	// "Suites" field
+	var suites = fmt.Sprintf("Suites: %s\n", aptConfig.Suites)
+	// "Components" field
+	var components = fmt.Sprintf("Components: %s\n", aptConfig.Components)
+	// "Architectures" field
+	var architecture = fmt.Sprintf("Architectures: %s\n", aptConfig.Arch)
+
+	//Download the official source signature and copy it from curlImage
+	aptSign, signPath := g.copyAPTSignature(root, aptConfig.Name, sign)
+	// "Signed-By" field
+	var signature = fmt.Sprintf("Signed-By: %s\n", signPath)
+
+	// Concatenate all fields together to be written to the same .sources file
+	var content strings.Builder
+	content.WriteString(enabled)
+	content.WriteString(types)
+	content.WriteString(uris)
+	content.WriteString(suites)
+	content.WriteString(components)
+	content.WriteString(signature)
+	content.WriteString(architecture)
+
+	return aptSign, content.String()
+}
+
+func (g generalGraph) compileRLang(root llb.State) llb.State {
+
+	var sign = "https://cloud.r-project.org/bin/linux/ubuntu/marutter_pubkey.asc"
+	var aptConfig = ir.APTConfig{
+		Name:       "R-base",
+		Enabled:    "yes",
+		Types:      "deb",
+		URIs:       "https://cloud.r-project.org/bin/linux/ubuntu",
+		Suites:     "focal-cran40/",
+		Components: "",
+		Signed:     "R-base.asc",
+		Arch:       "",
+	}
+
+	var file = fmt.Sprintf("/etc/apt/sources.list.d/%s.sources", aptConfig.Name)
+
+	var aptSource llb.State
+	var content string
+	aptSource, content = g.configRSrc(root, aptConfig, sign)
+
+	aptRLang := aptSource.
+		File(llb.Mkdir("/etc/apt/sources.list.d/", 0755, llb.WithParents(true)),
+			llb.WithCustomName("[internal] setting apt-source folder sources.list.d")).
+		File(llb.Mkfile(file, 0644, []byte(content)),
+			llb.WithCustomName("[internal] setting apt-source file")).
+		File(llb.Mkfile("/etc/apt/apt.conf.d/DEB822.conf", 0644, []byte("")),
+			llb.WithCustomName("[internal] setting apt-conf file to support DEB822 format")).
+		Run(llb.Shlex(fmt.Sprintf("bash -c \"%s\"", "echo 'APT::Sources::Use-Deb822 true;\n' > /etc/apt/apt.conf.d/DEB822.conf")),
+			llb.WithCustomName("[internal] enabled Deb822 format apt-source file"))
```

Comment:
Why not write the content into the file directly?

Reply:
Directly write to file when creating